plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageChecksPublisher.java

```diff
@@ -0,0 +1,426 @@
+     *
+     * @param listener
+     *         The task listener
+     */
+    void publishCoverageReport(final TaskListener listener) {
+        var publisher = ChecksPublisherFactory.fromRun(action.getOwner(), listener);
+        publisher.publish(extractChecksDetails());
+    }
+
+    @VisibleForTesting
+    ChecksDetails extractChecksDetails() {
+        var output = new ChecksOutputBuilder()
+                .withTitle(getChecksTitle())
+                .withSummary(getSummary())
+                .withAnnotations(getAnnotations())
+                .build();
+
+        return new ChecksDetailsBuilder()
+                .withName(checksName)
+                .withStatus(ChecksStatus.COMPLETED)
+                .withConclusion(getCheckConclusion(action.getQualityGateResult().getOverallStatus()))
+                .withDetailsURL(getCoverageReportBaseUrl())
+                .withOutput(output)
+                .build();
+    }
+
+    private String getChecksTitle() {
+        return String.format("%s: %s",
+                FORMATTER.getDisplayName(Baseline.MODIFIED_LINES),
+                action.formatValue(Baseline.MODIFIED_LINES, Metric.LINE));
+    }
+
+    private String getSummary() {
+        var root = action.getResult();
+        return getOverallCoverageSummary(root) + "\n\n"
+                + getQualityGatesSummary() + "\n\n"
+                + getProjectMetricsSummary(root);
+    }
+
+    private List<ChecksAnnotation> getAnnotations() {
+        if (annotationScope == ChecksAnnotationScope.SKIP) {
+            return List.of();
+        }
+        // TODO: check if it makes sense to add annotations for all lines
+        var annotations = new ArrayList<ChecksAnnotation>();
+        for (var fileNode : action.getResult().filterByModifiedLines().getAllFileNodes()) {
+            for (var aggregatedLines : getAggregatedMissingLines(fileNode)) {
+                ChecksAnnotationBuilder builder = new ChecksAnnotationBuilder()
+                        .withPath(fileNode.getPath())
+                        .withTitle(Messages.Checks_Annotation_Title())
+                        .withAnnotationLevel(ChecksAnnotationLevel.WARNING)
+                        .withMessage(getAnnotationMessage(aggregatedLines))
+                        .withStartLine(aggregatedLines.startLine)
+                        .withEndLine(aggregatedLines.endLine);
+
+                annotations.add(builder.build());
+            }
+        }
+
+        return annotations;
+    }
+
+    private String getAnnotationMessage(final AggregatedMissingLines aggregatedMissingLines) {
+        if (aggregatedMissingLines.startLine == aggregatedMissingLines.endLine) {
+            return Messages.Checks_Annotation_Message_SingleLine(aggregatedMissingLines.startLine);
+        }
+        return Messages.Checks_Annotation_Message_MultiLine(aggregatedMissingLines.startLine,
+                aggregatedMissingLines.endLine);
```

Comment:
Shouldn't the annotations query for the coverage metric as well? So we have different annotations depending on the metric? And where do you check if the line is covered or not? Should you evaluate the coverage per line as well? So I would expect something like:

- line 23 has no coverage
- line 32 has one missing branch
- line 45 has 4 mutations survived

I'll have a look how I can integrate such kind of method in the model.

I think the best approach would be to have a method in `FileNode` that returns these elements:

```
getMissedLines(final Metric metric)
```

Reply:
I added 3 methods now in https://github.com/uhafner/coverage-model/blob/b3b30707bce09541f2aa976d1a67177ca191c43b/src/main/java/edu/hm/hafner/metric/FileNode.java#L415-L450